zcash_client_backend/src/scan.rs

```diff
@@ -8,36 +8,107 @@ use std::sync::{
 };
 
 use memuse::DynamicUsage;
-use zcash_note_encryption::{batch, BatchDomain, Domain, ShieldedOutput, COMPACT_NOTE_SIZE};
+use zcash_note_encryption::{
+    batch, BatchDomain, Domain, ShieldedOutput, COMPACT_NOTE_SIZE, ENC_CIPHERTEXT_SIZE,
+};
 use zcash_primitives::{block::BlockHash, transaction::TxId};
 
-/// A decrypted note.
-pub(crate) struct DecryptedNote<A, D: Domain> {
+/// A decrypted transaction output.
+pub(crate) struct DecryptedOutput<A, D: Domain, M> {
     /// The tag corresponding to the incoming viewing key used to decrypt the note.
     pub(crate) ivk_tag: A,
     /// The recipient of the note.
     pub(crate) recipient: D::Recipient,
     /// The note!
     pub(crate) note: D::Note,
+    /// The memo field, or `()` if this is a decrypted compact output.
+    pub(crate) memo: M,
 }
 
-impl<A, D: Domain> fmt::Debug for DecryptedNote<A, D>
+impl<A, D: Domain, M> fmt::Debug for DecryptedOutput<A, D, M>
 where
     A: fmt::Debug,
     D::IncomingViewingKey: fmt::Debug,
     D::Recipient: fmt::Debug,
     D::Note: fmt::Debug,
-    D::Memo: fmt::Debug,
+    M: fmt::Debug,
 {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("DecryptedNote")
+        f.debug_struct("DecryptedOutput")
             .field("ivk_tag", &self.ivk_tag)
             .field("recipient", &self.recipient)
             .field("note", &self.note)
+            .field("memo", &self.memo)
             .finish()
     }
 }
 
+/// A decryptor of transaction outputs.
+pub(crate) trait Decryptor<A, D: BatchDomain, Output> {
+    type Memo;
+
+    // Once we reach MSRV 1.75.0, this can return `impl Iterator`.
+    fn batch_decrypt(
+        tags: &[A],
+        ivks: &[D::IncomingViewingKey],
+        outputs: &[(D, Output)],
+    ) -> Vec<Option<DecryptedOutput<A, D, Self::Memo>>>;
+}
+
+/// A decryptor of outputs as encoded in transactions.
+pub(crate) struct FullDecryptor;
```

Comment:
If this type is only pub(crate), doesn't that limit its utility somewhat? Same question about the trait. It's fine for now, but I expect eventually we'll want to expose them?

Reply:
Yep, the intention is that `BatchRunner` will become public, along with everything that is required for that. But I want to ensure that what is exposed is actually usable first. I've already made some adjustments locally while testing this with `zcashd`.